system logs ui changes

vishalmishraa opened the pull request:
## Linked Issue(s)

issue #555 

<!-- Link the issues being closed by or related to this PR. For example, you can use Closes #594 if this PR closes issue number 594 -->

## Acceptance Criteria fulfillment

<!-- This will contain the acceptance criteria required by the Pull Request in order to close the issue. This Section can be deleted if no acceptance criteria are provided. -->

- [ ] Task 1
- [ ] Task 2
- [ ] Task 3

## Proposed changes (including videos or screenshots)

![Screenshot 2024-10-05 at 1 53 58 AM](https://github.com/user-attachments/assets/1f228403-a029-4e18-807e-4ae6636317a2)
![Screenshot 2024-10-05 at 1 50 18 AM](https://github.com/user-attachments/assets/64bb9de3-20d5-48a9-8a9e-c56cab22f812)
![Screenshot 2024-10-05 at 1 50 03 AM](https://github.com/user-attachments/assets/511897b0-b924-453b-a099-d97209887665)
![Screenshot 2024-10-05 at 1 50 11 AM](https://github.com/user-attachments/assets/4eccda58-7095-49cc-86b9-85d17f8ec739)



<!-- CHANGELOG -->
<!--
  Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.
  If it fixes a bug or resolves a feature request, be sure to link to that issue below.
  This description will appear in the release notes if we accept the contribution.
-->

<!-- END CHANGELOG -->


## Further comments

<!-- If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc... -->

jayantbh commented:
Thanks for your contribution, @vishalmishraa!
Based on the screenshots it's certainly a decent improvement.

I would love to accept this PR, but it needs a few improvements to avoid performance issues.

Given that there could be thousands of lines being rendered, we do need to keep an eye out for excessive calls to JSON.stringify, Array.includes, and various other places.

Additionally, I think the metadata doesn't really need to be printed. What do you think?

jayantbh commented:
Linter failed. We need to sort out the linting issues but for now you can check our contributing guide to see how to fix that. Let us know if it doesn't work out for you.

vishalmishraa commented:
> Thanks for your contribution, @vishalmishraa! Based on the screenshots it's certainly a decent improvement.
> 
> I would love to accept this PR, but it needs a few improvements to avoid performance issues.
> 
> Given that there could be thousands of lines being rendered, we do need to keep an eye out for excessive calls to JSON.stringify, Array.includes, and various other places.
> 
> Additionally, I think the metadata doesn't really need to be printed. What do you think?

hi , @jayantbh , 

I removed the metadata entirely, so there’s no need to use JSON.stringify. and for  Array.includes, I’ll use Set(), as it offers O(1) time complexity, which is better. yup , i need to fix the linting issues.

jayantbh commented:
Directionally I think yes.
Your last commit looks better too.

I haven't done a proper review but I'll ask the team to check it out.

jayantbh commented:
This PR could use some tests. Especially with various kinds of logs that might come through.

jayantbh commented:
And of course, to test the performance of how many lines you're able to process via your functions as a benchmark.

jayantbh commented:
I ran a quick test myself based on the number of lines, and I think we'll hit other performance limits before we hit perf issues with the regexes and matches.

vishalmishraa commented:
> I ran a quick test myself based on the number of lines, and I think we'll hit other performance limits before we hit perf issues with the regexes and matches.

okay , i think i need to change my approach ,Could you suggest an alternatives or any way to do that ? I will work on implementing it accordingly.

jayantbh commented:
No no. I'm saying you might not need to change your approach.

vishalmishraa commented:
so , do i need to do any improvements or it is good to go ?

jayantbh commented:
> This PR could use some tests. Especially with various kinds of logs that might come through.

This.

vishalmishraa commented:
Okay, I’m working on it. I’ll be writing tests in ./web-server/src/utils/test, where I’ll create a logFormatter.test.ts file and write the tests. I need some help with generating different types of system logs. Should I generate new logs for testing, or can I use the logs that are being generated right now?

vishalmishraa commented:
@jayantbh  I need help regarding generating different types of system logs. Should I generate new logs for testing? If yes, how can I generate other different logs? Or is it better to use the logs that are currently being generated?

jayantbh commented:
I think handle for the variety of logs that are generated today.
But ensure that when and if something unexpected comes in the logs (such as, missing HTTP method in HTTP logs, dates formatted in a different way, etc.) Then it doesn't break the UI.

vishalmishraa commented:
i think the function I wrote is not feasible for parsing all types of logs. There are several different types of logs, such as system-generated logs (some print statements in backend code), HTTP server error logs, and others.

i did some research on better logging solutions and found a CNCF-based library called [Fluentd](https://www.fluentd.org/), which offers unified logging, better parsing, and improved data collection and observation. It's also lightweight, requiring only 40 to 50 MB of space. I think it could be a good fit for this project.

more details:
- [What is Fluentd?](https://www.fluentd.org/architecture)
- [How Fluentd works (YouTube video)](https://youtu.be/5ofsNyHZwWE?t=257)

jayantbh commented:
I suspected it could go in this way.
You can go with a simpler approach altogether as well.

jayantbh commented:
> > I suspected it could go in this way. You can go with a simpler approach altogether as well.
> 
> Then , there will be some logs, which can't be parsed, and they will be displayed as it is without colors. I'll try to add regex for logs that are frequently generated, should i proceed with this ?

Yea, I think you could try that.

Something is better than nothing after all. But make sure you have tests.

vishalmishraa commented:
> > > I suspected it could go in this way. You can go with a simpler approach altogether as well.
> > 
> > 
> > Then , there will be some logs, which can't be parsed, and they will be displayed as it is without colors. I'll try to add regex for logs that are frequently generated, should i proceed with this ?
> 
> Yea, I think you could try that.
> 
> Something is better than nothing after all. But make sure you have tests.

Thanks, @jayantbh. I've updated the functions and added the tests. Please review them.

jayantbh commented:
Hmm, not sure I feel great about that approach.
How did you try to add and test the error boundary that it caused a higher level boundary to catch it instead?

vishalmishraa commented:
Hi, @jayantbh,  can you please check this now , as i've implemented error handling . also now i'm not using another component , combined it in SystemLog.tsx , with error message .

vishalmishraa commented:
> Hmm, not sure I feel great about that approach. How did you try to add and test the error boundary that it caused a higher level boundary to catch it instead?

i did this : 



```

const SystemLogs = () => {
    return (
        <ErrorBoundary FallbackComponent={SystemLogErrorFallback}>
            // code
        </ErrorBoundary>
    );
};

```

But it didn’t work, and I think it should have worked.

vishalmishraa commented:
Also i think the approach I'm using now would be work good  because the component which should be  passed into the ErrorBoundary fallback should also receive serviceName and other props as it should render logs in plain text. So, in the future, if we make any improvements to the UI for SystemLogs like (implementing search feature ,etc..), we would need to apply the same modifications to the custom fallback component as well, since it's rendering system logs in plain text. Therefore, it would be better to handle this in the same position .

vishalmishraa commented:
Hi @jayantbh ,

I’ve implemented the error boundary and apologize for the confusion earlier. You were right , I was testing it incorrectly, which is why it didn’t work. However, I’ve now implemented it properly. I’ve also created a useSystemLog hook for it and separated the component.Please check it now.

jayantbh commented:
@vishalmishraa could you share screenshots of the error boundary in effect?

vishalmishraa commented:
> @vishalmishraa could you share screenshots of the error boundary in effect?

### Sure : 


https://github.com/user-attachments/assets/e05ff22a-ad98-474b-892f-105bd3b3eef6

jayantbh commented:
Looks great!
The linter appears to have failed. Do check what that's about.

Meanwhile, we'll give this a round of final reviews. But I think we're looking good visually/functionally.

vishalmishraa commented:
i think linter is failing because of workflow 

![Screenshot 2024-10-15 at 2 15 56 PM](https://github.com/user-attachments/assets/8dcaed5c-9859-4b89-a3f6-decf7fed55d0)

jayantbh commented:
Okay, I guess ignore that for now. We'll do the reviews as soon as we're able. :)

e-for-eshaan commented on web-server/src/utils/logFormatter.ts:
```diff
@@ -0,0 +1,103 @@
+export const parseLogLine = (rawLogLine: string): ParsedLog | null => {
+  const generalLogMatch = rawLogLine.match(generalLogRegex);
+  if (generalLogMatch) {
+    const [, timestamp, , logLevel, message] = generalLogMatch;
```
```ts
const [_pickAname, timestamp, _pickAnotherName, logLevel, message]
```

vishalmishraa replied:
got it.

e-for-eshaan commented on web-server/src/utils/logFormatter.ts:
```diff
@@ -0,0 +1,103 @@
+    const [method, path] = request.split(' ');
+    return {
+      timestamp,
+      logLevel: 'INFO', // Assuming all HTTP logs are INFO level
```
all these string constants should be wrapped in an enum in our `src/types/resources.ts`
```ts
enum YourEnumName 
{ 
  INFO = 'INFO', 
  DEBUG = 'DEBUG'
  ...other_possible_values... 
}
```

vishalmishraa replied:
done , created a enum LogLevel.

e-for-eshaan commented on web-server/src/utils/logFormatter.ts:
```diff
@@ -0,0 +1,103 @@
+
+  const postgresLogMatch = rawLogLine.match(postgresLogRegex);
+  if (postgresLogMatch) {
+    const [, timestamp, , logLevel, message] = postgresLogMatch;
```
same:
```ts
const [_pickAname, timestamp, _pickAnotherName, logLevel, message]
```

e-for-eshaan commented on web-server/src/utils/logFormatter.ts:
```diff
@@ -0,0 +1,103 @@
+
+  const dataSyncLogMatch = rawLogLine.match(dataSyncLogRegex);
+  if (dataSyncLogMatch) {
+    const [, logLevel, action, service, message] = dataSyncLogMatch;
```
same

e-for-eshaan commented on web-server/src/constants/log-formatter.ts:
```diff
@@ -0,0 +1,31 @@
+export const postgresMultiLineLogRegex =
+  /^(?<timestamp>\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}\.\d{3} UTC) \[\d+\] (?<loglevel>[A-Z]+):\s*(?<message>(.|\n)+?)$/;
+export const dataSyncLogRegex = /\[(\w+)\]\s(.+?)\sfor\s(\w+)\s(.+)/;
+export const validLogLevels = new Set([
+  'DEBUG',
+  'INFO',
+  'NOTICE',
+  'WARNING',
+  'ERROR',
+  'LOG',
+  'FATAL',
+  'PANIC',
+  'STATEMENT',
+  'DETAIL'
+]);
```
this should be an enum in our `resources` file

vishalmishraa replied:
I added these as enums and also removed the validLogLevels logic from the parser. Previously, I was using two approaches to parse PostgreSQL logs, but during testing, I noticed that my first approach, PostgresMultilineLogMatch, can parse both types of PostgreSQL logs. Therefore, there's no need to write separate logic for each type ,refer to the [official postgres documentation  ](https://www.postgresql.org/docs/current/runtime-config-logging.html#:~:text=server%20command%20line.-,log_line_prefix,-(string)). and  [this article ](https://betterstack.com/community/guides/logging/how-to-start-logging-with-postgresql/#structured-logging-in-json:~:text=%25m%20%5B%25p%5D%20%25q,the%20following%20details%3A) also for  [log levels ](https://www.postgresql.org/docs/current/runtime-config-logging.html#:~:text=Table%C2%A019.2.-,Message%20Severity%20Levels,-Severity).

e-for-eshaan commented on web-server/src/utils/logFormatter.ts:
```diff
@@ -0,0 +1,95 @@
+export const parseLogLine = (rawLogLine: string): ParsedLog | null => {
+  const generalLogMatch = rawLogLine.match(generalLogRegex);
+  if (generalLogMatch) {
+    const [_fullLog, timestamp, _unused, logLevel, message] = generalLogMatch;
```
`_unused`
this looks off, name it correctly to what this variable stands for, and the underscore is enough to know that this variable is unused

vishalmishraa replied:
done. ,changing it with _pid

e-for-eshaan commented on web-server/src/utils/logFormatter.ts:
```diff
@@ -0,0 +1,95 @@
+    const [
+      _fullLog,
+      ip,
+      _unused,
+      _unused2,
```
same with these :)

vishalmishraa replied:
changed the second one to _username , and first one with _ (underscore) [refer to this](https://docs.gunicorn.org/en/stable/settings.html#:~:text=remote%20address-,l,%27%2D%27,-u)

e-for-eshaan reviewed (dismissed):
LGTM, good stuff @vishalmishraa 🎉

shivam-bit commented on web-server/src/components/Service/SystemLog/FormattedLog.tsx:
```diff
@@ -0,0 +1,63 @@
+
+  const { timestamp, ip, logLevel, message } = log;
+  return (
+    <Line
+      key={index}
+      marginBottom={'8px'}
+      fontSize={'14px'}
+      fontFamily={'monospace'}
```
same as other requested changes

shivam-bit commented on web-server/src/components/Service/SystemLog/FormattedLog.tsx:
```diff
@@ -0,0 +1,63 @@
+      fontSize={'14px'}
+      fontFamily={'monospace'}
+    >
+      <Line component="span" color="info.main">
```
i think you can do this like `<Line  color="info">`

vishalmishraa commented on web-server/src/constants/log-formatter.ts:
```diff
@@ -0,0 +1,17 @@
+export interface ParsedLog {
+  timestamp: string;
+  logLevel: string;
+  message: string;
+  role?: string;
+  ip?: string;
+}
```
done

shivam-bit commented on web-server/src/components/Service/SystemLog/FormattedLog.tsx:
```diff
@@ -0,0 +1,52 @@
+
+  const { timestamp, ip, logLevel, message } = log;
+  return (
+    <Line medium marginBottom={'8px'} fontFamily={'monospace'}>
```
Try this`<Line mono marginBottom={1}>`

shivam-bit commented on web-server/src/components/Service/SystemLog/PlainLog.tsx:
```diff
@@ -0,0 +1,9 @@
+
+export const PlainLog = ({ log }: { log: string; index: number }) => {
+  return (
+    <Line medium marginBottom={'8px'} fontFamily={'monospace'}>
```
Try this `<Line mono marginBottom={1}>`, it should ideally give you the same effect.

vishalmishraa replied:
done

shivam-bit commented:
@vishalmishraa LGTM!🚀🚀🚀

jayantbh commented:
I think this doesn't need another round of reviews. We can go ahead and merge this. Any remaining changes will probably be minor and can be taken up in subsequent PRs.

jayantbh commented:
Congrats on getting this merged!
Good work my dude!
